orchestrator/health/health.go

```diff
@@ -53,69 +67,101 @@ func (monitorObj *Monitor) ProcessList() map[int]process.Process {
-func (monitorObj *Monitor) listenToHealthChecks(wg *sync.WaitGroup) {
-	defer wg.Done()
+func (monitorObj *Monitor) listenToHealthChecks() {
+	defer monitorObj.wg.Done()
 	defer monitorObj.connectionHandler.Close()
 
 	monitorObj.establishConnection()
+	log.Println(logPrefix, "Listening for health checks")
+
+	var updateLastSeenWG sync.WaitGroup
 
 	for {
-		healthCheck, _ := monitorObj.connectionHandler.RecvString(zmq4.DONTWAIT)
+		select {
+		case <-monitorObj.shutdown:
+			log.Println(logPrefix, "Health checker is shutting down")
+			updateLastSeenWG.Wait()
 
-		if healthCheck != "" {
-			log.Println("Received", healthCheck)
+			return
+		default:
+			healthCheck, _ := monitorObj.connectionHandler.RecvString(zmq4.DONTWAIT)
 
-			go monitorObj.updateProcessLastSeen(healthCheck)
+			if healthCheck != "" {
+				log.Println(logPrefix, "Received", healthCheck)
+
+				updateLastSeenWG.Add(1)
+				go monitorObj.updateProcessLastSeen(healthCheck, &updateLastSeenWG)
+			}
 		}
 	}
 }
 
 // trackLiveness A function to track the alive status of processes in the process list
-func (monitorObj *Monitor) trackLiveness(wg *sync.WaitGroup) {
-	defer wg.Done()
+func (monitorObj *Monitor) trackLiveness() {
+	defer monitorObj.wg.Done()
+
+	log.Println(logPrefix, "Tracking processes liveness")
 
 	for {
-		monitorObj.processListMutex.Lock()
+		select {
+		case <-monitorObj.shutdown:
+			log.Println(logPrefix, "Liveness tracker is shutting down")
+			return
+		default:
+			monitorObj.processListMutex.Lock()
 
-		for id, process := range monitorObj.processList {
-			if process.Trackable {
-				delay := time.Now().Sub(process.LastPing)
+			for id, process := range monitorObj.processList {
+				if process.Trackable {
+					delay := time.Now().Sub(process.LastPing)
 
-				if delay > monitorObj.livenessProbe {
-					process.Trackable = false
-					monitorObj.processList[id] = process
+					if delay > monitorObj.livenessProbe {
+						process.Trackable = false
+						monitorObj.processList[id] = process
 
-					log.Println(fmt.Sprintf("Process with id: %d went offline", id))
+						log.Println(logPrefix, fmt.Sprintf("Process with id: %d went offline", id))
 
-					//TODO: Add logic to kill that process and spawn another
+						//TODO: Add logic to kill that process and spawn another
+					}
 				}
 			}
-		}
 
-		monitorObj.processListMutex.Unlock()
+			monitorObj.processListMutex.Unlock()
+		}
 	}
 }
 
 // updateProcessLastSeen A function to update the last seen of processes in the process list
-func (monitorObj *Monitor) updateProcessLastSeen(healthCheck string) {
+func (monitorObj *Monitor) updateProcessLastSeen(healthCheck string, wg *sync.WaitGroup) {
+	defer wg.Done()
+
 	processUtil, err := process.ParseUtilization(healthCheck)
-	if err {
-		log.Println("Error unmarshalling health check ping")
-	}
+	errors.HandleError(err, fmt.Sprintf("%s%s", logPrefix, "Unable to unmarshal health check ping at updateProcessLastSeen()"), false)
```

Comment:
if there is an error, the next line will explode anyway and panic the process since ```processUtil``` is not parsed so ```processUtil.PID```. You can just ignore the message if it's not parsed.

Reply:
Nice catch.

Reply:
I checked, it actually wouldn't crash because if this line `processUtil, err := process.ParseUtilization(healthCheck)` encountered a parsing error, it will return an empty `Utilization{}` and not nil.